daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala

```diff
@@ -0,0 +1,884 @@
+import com.daml.lf.value.Value.{ContractId, ContractInstance}
+import org.scalatest.{Assertion, Inside}
+import org.scalatest.freespec.AnyFreeSpec
+import org.scalatest.matchers.should.Matchers
+import com.daml.lf.speedy.SBuiltin.{SBFetchAny, SBUFetchKey, SBULookupKey}
+import com.daml.lf.speedy.SValue.SContractId
+import com.daml.lf.transaction.{GlobalKey, GlobalKeyWithMaintainers, TransactionVersion, Versioned}
+import com.daml.lf.testing.parser.Implicits._
+import org.scalatest.matchers.{MatchResult, Matcher}
+
+class ExplicitDisclosureTest extends ExplicitDisclosureTestMethods {
+
+  import ExplicitDisclosureTest._
+
+  "disclosed contract behaviour" - {
+    "fetching contracts" - {
+      "test data validation" in {
+        ledgerParty should not be disclosureParty
+        ledgerParty should not be maintainerParty
+        disclosureParty should not be maintainerParty
+        getOwner(ledgerCaveContract.unversioned.arg) shouldBe Some(ledgerParty)
+        disclosedCaveContract.contractId shouldBe SContractId(contractId)
+        getOwner(disclosedCaveContract.argument.toUnnormalizedValue) shouldBe Some(disclosureParty)
+      }
+
+      "disclosure preprocessing" - {
+        "template does not exist" in {
+          templateDoesNotExist(
+            SBFetchAny(SEValue(SContractId(contractId)), SEValue.None),
+            disclosedCaveContractInvalidTemplate,
+          )
+        }
+      }
+
+      "ledger queried when contract ID is not disclosed" in {
+        ledgerQueriedWhenContractNotDisclosed(
+          SBFetchAny(SEValue(SContractId(contractId)), SEValue.None),
+          contractId,
+          getContract = Map(contractId -> ledgerCaveContract),
+        )(result =>
+          inside(result) {
+            case Right(SValue.SAny(_, contract @ SValue.SRecord(`caveTemplateId`, _, _))) =>
+              getOwner(contract.toUnnormalizedValue) shouldBe Some(ledgerParty)
+          }
+        )
+      }
+
+      "disclosure table queried when contract ID is disclosed" - {
+        "contract ID in disclosure table only" in {
+          disclosureTableQueriedWhenContractDisclosed(
+            SBFetchAny(SEValue(SContractId(contractId)), SEValue.None),
+            disclosedCaveContract,
+            disclosedContracts = ImmArray(disclosedCaveContract),
+          )(result =>
+            inside(result) {
+              case Right(SValue.SAny(_, contract @ SValue.SRecord(`caveTemplateId`, _, _))) =>
+                getOwner(contract.toUnnormalizedValue) shouldBe Some(disclosureParty)
+            }
+          )
+        }
+
+        "contract ID in ledger and disclosure table" in {
+          disclosureTableQueriedWhenContractDisclosed(
+            SBFetchAny(SEValue(SContractId(contractId)), SEValue.None),
+            disclosedCaveContract,
+            // contract owned by ledgerParty
+            getContract = Map(contractId -> ledgerCaveContract),
+            // contract owned by disclosureParty
```

Comment:
I don't quite understand why you need different owners here.

Reply:
I'll remove those comments as I don't think they're adding any value here.

Each contract is distinguished by its owner (c.f. boron tagging) and is associated with a common contract ID (via the ledger and the disclosure table). So, when we look at the contract returned from Speedy, looking at the contract's owner field then tells us definitively what source was used to supply the contract (i.e. was it the ledger or the disclosure table). 
Here, observing the owner informs us that we expect to serve the contract from the disclosure table and not the ledger.